internal/controller/istio_token_ratelimit_cluster_reconciler.go

```diff
@@ -0,0 +1,203 @@
+				labels.Set(child.(*controller.RuntimeObject).GetLabels()).AsSelector().Matches(labels.Set(desiredEnvoyFilter.GetLabels()))
+		})
+
+		// create
+		if !found {
+			modifiedGateways = append(modifiedGateways, gateway.GetLocator()) // we only signal the gateway as modified when an envoyfilter is created, because updates won't change the status
+			desiredEnvoyFilterUnstructured, err := controller.Destruct(desiredEnvoyFilter)
+			if err != nil {
+				logger.Error(err, "failed to destruct envoyfilter object", "gateway", gatewayKey.String(), "envoyfilter", desiredEnvoyFilter)
+				continue
+			}
+			if _, err = resource.Create(ctx, desiredEnvoyFilterUnstructured, metav1.CreateOptions{}); err != nil {
+				logger.Error(err, "failed to create envoyfilter object", "gateway", gatewayKey.String(), "envoyfilter", desiredEnvoyFilterUnstructured.Object)
+				// TODO: handle error
+			}
+			continue
+		}
+
+		existingEnvoyFilter := existingEnvoyFilterObj.(*controller.RuntimeObject).Object.(*istioclientgonetworkingv1alpha3.EnvoyFilter)
+
+		if kuadrantistio.EqualEnvoyFilters(existingEnvoyFilter, desiredEnvoyFilter) {
+			logger.V(1).Info("envoyfilter object is up to date, nothing to do")
+			continue
+		}
+
+		// update
+		existingEnvoyFilter.Spec = istioapinetworkingv1alpha3.EnvoyFilter{
+			TargetRefs:    desiredEnvoyFilter.Spec.TargetRefs,
+			ConfigPatches: desiredEnvoyFilter.Spec.ConfigPatches,
+			Priority:      desiredEnvoyFilter.Spec.Priority,
+		}
+
+		existingEnvoyFilterUnstructured, err := controller.Destruct(existingEnvoyFilter)
+		if err != nil {
+			logger.Error(err, "failed to destruct envoyfilter object", "gateway", gatewayKey.String(), "envoyfilter", existingEnvoyFilter)
+			continue
+		}
+		if _, err = resource.Update(ctx, existingEnvoyFilterUnstructured, metav1.UpdateOptions{}); err != nil {
+			logger.Error(err, "failed to update envoyfilter object", "gateway", gatewayKey.String(), "envoyfilter", existingEnvoyFilterUnstructured.Object)
+			// TODO: handle error
+		}
+	}
+
+	state.Store(StateIstioTokenRateLimitClustersModified, modifiedGateways)
+
+	// cleanup istio clusters for gateways that are not in the effective policies
+	staleEnvoyFilters := topology.Objects().Items(func(o machinery.Object) bool {
+		_, desired := desiredEnvoyFilters[k8stypes.NamespacedName{Name: o.GetName(), Namespace: o.GetNamespace()}]
+		return o.GroupVersionKind().GroupKind() == kuadrantistio.EnvoyFilterGroupKind &&
+			labels.Set(o.(*controller.RuntimeObject).GetLabels()).AsSelector().Matches(TokenRateLimitObjectLabels()) &&
+			!desired
+	})
+	for _, envoyFilter := range staleEnvoyFilters {
+		if err := r.client.Resource(kuadrantistio.EnvoyFiltersResource).Namespace(envoyFilter.GetNamespace()).Delete(ctx, envoyFilter.GetName(), metav1.DeleteOptions{}); err != nil {
+			logger.Error(err, "failed to delete envoyfilter object", "envoyfilter", fmt.Sprintf("%s/%s", envoyFilter.GetNamespace(), envoyFilter.GetName()))
+			// TODO: handle error
```

Comment:
Should we implement this todo before merging ?

Reply:
same